From 9ecea4ba3b01faa867e22c5b01c4fbdcdb20bb45 Mon Sep 17 00:00:00 2001
From: Jerome Martinez <jerome@mediaarea.net>
Date: Thu, 4 Jul 2024 10:45:57 +0200
Subject: [PATCH 04/18] avcodec/dpx: Fix B&W film scans from DIAMANT-Film

---
 libavcodec/dpx.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/libavcodec/dpx.c b/libavcodec/dpx.c
index 52a03501ec..b9a3e1d757 100644
--- a/libavcodec/dpx.c
+++ b/libavcodec/dpx.c
@@ -615,6 +615,7 @@ static int decode_frame(AVCodecContext *avctx, AVFrame *p,
     // Some devices do not pad 10bit samples to whole 32bit words per row
     dpx->unpadded_10bit = 0;
     if (!memcmp(input_device, "Scanity", 7) ||
+        !memcmp(creator, "DIAMANT-Film", 12) ||
         !memcmp(creator, "Lasergraphics Inc.", 18)) {
         if (avctx->bits_per_raw_sample == 10)
             dpx->unpadded_10bit = 1;
-- 
2.52.0

