From 8e3f6553a97a1820d794f11555a0add5cb0e2803 Mon Sep 17 00:00:00 2001
From: Jerome Martinez <jerome@mediaarea.net>
Date: Tue, 2 Dec 2025 12:29:03 +0100
Subject: [PATCH 05/18] configure evaluate toolchain to determine suffix

---
 configure | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/configure b/configure
index b09e9d0436..032266f3c2 100755
--- a/configure
+++ b/configure
@@ -4752,11 +4752,15 @@ fi
 exesuf() {
     case $1 in
         mingw32*|mingw64*|msys*|win32|win64|cygwin*|*-dos|freedos|opendos|os/2*|symbian) echo .exe ;;
+        *)
+            case $2 in
+                msvc) echo .exe ;;
+            esac ;;
     esac
 }
 
-EXESUF=$(exesuf $target_os)
-HOSTEXESUF=$(exesuf $host_os)
+EXESUF=$(exesuf $target_os $toolchain)
+HOSTEXESUF=$(exesuf $host_os $toolchain)
 
 # set temporary file name
 : ${TMPDIR:=$TEMPDIR}
-- 
2.52.0

